[MM-35005]: Always execute a wildcard search when searching for files

kyeongsoosoo opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->

When searching for files, wildcard postfix is added to `term` on the backend.
Wildcard searches are executed regardless of the user entering "*" in the search term.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX
Otherwise, link the JIRA ticket.
-->
Fixes https://github.com/mattermost/mattermost-server/issues/21189
#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Always execute a wildcard search when searching for files
```

mattermod commented:
Hello @kyeongsoosoo,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

kyeongsoosoo commented:
I will re-open it after the tests are fixed!

kyeongsoosoo commented on store/searchtest/file_info_layer.go:
```diff
@@ -1214,7 +1256,7 @@ func testFileInfoSearchDiscardWildcardAlone(t *testing.T, th *SearchTestHelper)
 	p1, err := th.createFileInfo(th.User.Id, post.Id, "qwerty", "qwerty", "jpg", "image/jpeg", 0, 0)
 	require.NoError(t, err)
-	_, err = th.createFileInfo(th.User.Id, post.Id, "qwertyjkl", "qwertyjkl", "jpg", "image/jpeg", 0, 0)
+	_, err = th.createFileInfo(th.User.Id, post.Id, "werty", "werty", "jpg", "image/jpeg", 0, 0)
```
I changed this part to match the code change

kyeongsoosoo replied:
The purpose of this test is to show that wildcards cannot be used alone.
` "qwerty *"` is used as a term for testing but the test failed because the wildcard was added to every terms.
(`qwerty* matches qwertyjkl)

kyeongsoosoo commented:
@isacikgoz This PR is ready for review!
The remaining failing tests don't seem to be relevant to this PR.

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

isacikgoz commented:
@kyeongsoosoo re-running the tests, not sure if the failure was about your changes.

agnivade commented:
I'll let Alejandro review this since this has some performance implications we are facing right now. Also note that this also needs changes to elasticsearch code. Some staff member needs to do that.

agarciamontoro commented on store/sqlstore/file_info_store.go:
```diff
@@ -625,11 +624,32 @@ func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, team
 
+			exactPhraseTerms := []string{}
+
+			if exactPhraseRegExp, exactPhraseRegExpErr := regexp.Compile(`".+"\s*`); exactPhraseRegExpErr == nil {
```
Instead of compiling this in every iteration of the loop, we should store the regex as a variable using `regexp.MustCompile` outside of the function.

Also, this regular expression matches something like

agarciamontoro commented on services/searchengine/bleveengine/search.go:
```diff
@@ -708,7 +709,27 @@ func (b *BleveEngine) SearchFiles(channels model.ChannelList, searchParams []*mo
 		if params.Terms != "" {
 			terms := []string{}
-			for _, term := range strings.Split(params.Terms, " ") {
+			exactPhraseTerms := []string{}
+
+			if exactPhraseRegExp, exactPhraseRegExpErr := regexp.Compile(`".+"\s*`); exactPhraseRegExpErr == nil {
+				exactPhraseTerms = append(exactPhraseTerms, exactPhraseRegExp.FindAllString(params.Terms, -1)...)
+				params.Terms = exactPhraseRegExp.ReplaceAllLiteralString(params.Terms, "")
+			}
+
+			wildcardAddedTerms := strings.Fields(params.Terms)
+			wildcardRegExp, regExpErr := regexp.Compile(`\*?$`)
+
+			if regExpErr == nil {
+				for index, term := range wildcardAddedTerms {
+					if !strings.HasPrefix(term, "*") {
+						wildcardAddedTerms[index] = wildcardRegExp.ReplaceAllLiteralString(term, "*")
+					}
+				}
+			}
+
+			parsedTerms := append(wildcardAddedTerms, exactPhraseTerms...)
+
+			for _, term := range parsedTerms {
```
Also the same comments as with the DB Postgres case: regexp should be defined outside, just once, with MustCompile. And the first regexp ".+"\s*, should be fixed.

agarciamontoro commented on store/searchtest/file_info_layer.go:
```diff
@@ -1185,6 +1190,43 @@ func testFileInfoSupportWildcards(t *testing.T, th *SearchTestHelper) {
+	require.NoError(t, err)
+	defer th.deleteUserFileInfos(th.User.Id)
+
+	t.Run("supports wildcard search without wildcards", func(t *testing.T) {
+		params := &model.SearchParams{
+			Terms: "search",
+		}
+		results, err := th.Store.FileInfo().Search([]*model.SearchParams{params}, th.User.Id, th.Team.Id, 0, 20)
+		require.NoError(t, err)
+
+		require.Len(t, results.FileInfos, 2)
+		th.checkFileInfoInSearchResults(t, p1.Id, results.FileInfos)
+		th.checkFileInfoInSearchResults(t, p2.Id, results.FileInfos)
+	})
+
+	t.Run("support wildcard search with another term placed after without wildcards", func(t *testing.T) {
+		params := &model.SearchParams{
+			Terms: "sear post",
+		}
+		results, err := th.Store.FileInfo().Search([]*model.SearchParams{params}, th.User.Id, th.Team.Id, 0, 20)
+		require.NoError(t, err)
+
+		require.Len(t, results.FileInfos, 1)
+		th.checkFileInfoInSearchResults(t, p1.Id, results.FileInfos)
+	})
```
We should add some more tests here where `Terms` is something like `"one exact match" one two three "another exact match"`

kyeongsoosoo replied:
I added a test case for `exact phrases`

agarciamontoro reviewed (changes requested):
Thank for this, @kyeongsoosoo! :tada: 

I left some comments below that we need to address before going on, but performance-wise, I don't think this will have a negative impact: since the `tsvectors` are not changed, the indexes we were using are still in use. I'd still be interested in knowing what the actual performance hit would be after adding `*` to most terms, but we can wait for the release performance comparison to find that out.

kyeongsoosoo commented on store/sqlstore/file_info_store.go:
```diff
@@ -625,11 +624,32 @@ func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, team
 
+			exactPhraseTerms := []string{}
+
+			if exactPhraseRegExp, exactPhraseRegExpErr := regexp.Compile(`".+"\s*`); exactPhraseRegExpErr == nil {
```
I fixed the regular expression. Thanks for the suggestion. 
I removed the last part(`\s*`) because the spaces will be erased when we use `strings.Fields` in the next line.

kyeongsoosoo commented on store/sqlstore/file_info_store.go:
```diff
@@ -625,11 +624,32 @@ func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, team
+				for _, term := range exactPhraseRegExp.FindAllString(terms, -1) {
+					quoteRemovedTerm := strings.Trim(term, "\"")
+
+					exactPhraseTerms = append(exactPhraseTerms, strings.Fields(quoteRemovedTerm)...)
```
I thought we had to split it into single words to use `to_tsquery`.

kyeongsoosoo commented:
@agarciamontoro Thanks for the review! I fixed the issues :)

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

agarciamontoro reviewed (changes requested):
Ugh, sorry, I had reviewed this but forgot to click on Submit :facepalm: Sorry for the delay, my bad!

I've left a couple more comments to address. As a general comment, I think we should comment this code thoroughly, since it's quite complex and sometimes difficult to follow. Could you add comments explaining each step?

Also, I think we'll need a thorough QA review here, since just reading the code makes me feel I could be missing some stuff.

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

kyeongsoosoo commented:
@agarciamontoro  Thank you for a review. I've been busy lately, so I'll work on it this weekend

agarciamontoro commented:
Thanks, @kyeongsoosoo!

kyeongsoosoo commented:
@agarciamontoro I reflected the review and added comments. Please let me know if there is anything I need to add more comments

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

agarciamontoro commented on services/searchengine/bleveengine/search.go:
```diff
@@ -17,6 +18,13 @@ import (
 const DeletePostsBatchSize = 500
 const DeleteFilesBatchSize = 500
 
+// Find the end of the term and include the wildcard if it exists
+// because we will put a wildcard at the end of the term regardless of whether a wildcard exists or not
+var wildcardRegExpForFileSearch = regexp.MustCompile(`\*?$`)
+
+// In excludedTerms case, we don't put a wildcard after the term.
+var exactPhraseRegExpForFileSearch = regexp.MustCompile(`"[^"]+"`)
+
```
These comments can be modified to something more descriptive. Let's say:
```suggestion
// Regexp to match an asterisk at the end of the string. It is used
// in the file search logic to put a wildcard at the end of the search
// term regardless of whether a wildcard exists or not
var wildcardRegExpForFileSearch = regexp.MustCompile(`\*?$`)

// Regexp to mathch any quoted string. It is used in the file search
// logic to avoid adding a wildcard at the end of the quoted search
// terms, which need to be matched literally.
var exactPhraseRegExpForFileSearch = regexp.MustCompile(`"[^"]+"`)

```

agarciamontoro commented on services/searchengine/bleveengine/search.go:
```diff
@@ -708,7 +716,25 @@ func (b *BleveEngine) SearchFiles(channels model.ChannelList, searchParams []*mo
+
+			wildcardAddedTerms := strings.Fields(params.Terms)
+
+			// A wildcard is attached to the end of every word
+			// regardless of whether or not a wildcard exists
+			for index, term := range wildcardAddedTerms {
+				if !strings.HasPrefix(term, "*") {
+					wildcardAddedTerms[index] = wildcardRegExpForFileSearch.ReplaceAllLiteralString(term, "*")
+				}
+			}
```
Sorry, but I don't understand this: Why do we check if the term starts with an asterisk and skip in that case?

agarciamontoro replied:
Also, is this the same as the following?

```go
// A wildcard is attached to the end of every word
// regardless of whether or not a wildcard exists
for index, term := range wildcardAddedTerms {
	if !strings.HasPrefix(term, "*") && !strings.HasSuffix(term, "*") {
		wildcardAddedTerms[index] = term + "*"
	}
}
```

agarciamontoro commented on store/sqlstore/file_info_store.go:
```diff
@@ -44,6 +44,16 @@ type fileInfoWithChannelID struct {
 	Archived        bool
 }
 
+// Find the end of the term and include the wildcard if it exists
+// because we will put a wildcard at the end of the term regardless of whether a wildcard exists or not
+var wildcardRegExpForFileSearch = regexp.MustCompile(`\*?$`)
+
+// Find all characters between quotes
+var exactPhraseRegExpForFileSearch = regexp.MustCompile(`"[^"]+"`)
+
```
Similar comment as in the other file:
```suggestion
// Regexp to match an asterisk at the end of the string. It is used
// in the file search logic to put a wildcard at the end of the search
// term regardless of whether a wildcard exists or not
var wildcardRegExpForFileSearch = regexp.MustCompile(`\*?$`)

// Regexp to mathch any quoted string. It is used in the file search
// logic to avoid adding a wildcard at the end of the quoted search
// terms, which need to be matched literally.
var exactPhraseRegExpForFileSearch = regexp.MustCompile(`"[^"]+"`)

```

agarciamontoro commented on store/sqlstore/file_info_store.go:
```diff
@@ -44,6 +44,16 @@ type fileInfoWithChannelID struct {
+// Find all characters between quotes
+var exactPhraseRegExpForFileSearch = regexp.MustCompile(`"[^"]+"`)
+
+// In excludedTerms case, we don't put a wildcard after the term.
+var wildcardRegExpForPostgresExcludedTerms = regexp.MustCompile(`\*($| )`)
+
```
And some more rewording here, although I must admit I don't fully understand this regex.
```suggestion
// Regexp to match an asterisk, either at the end of the string, or right before a space.
// It is used in the file search logic so that we don't put a wildcard after the excluded terms.
var wildcardRegExpForPostgresExcludedTerms = regexp.MustCompile(`\*($| )`)

```

agarciamontoro commented on store/sqlstore/file_info_store.go:
```diff
@@ -614,22 +625,38 @@ func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, team
+
+			wildcardAddedTerms := strings.Fields(terms)
+
+			for index, term := range wildcardAddedTerms {
+				// A wildcard is attached to the end of every word
+				// regardless of whether or not a wildcard exists
+				wildcardAddedTerms[index] = wildcardRegExpForFileSearch.ReplaceAllLiteralString(term, ":*")
+			}
```
Similarly: is this the same as
```suggestion
			for index, term := range wildcardAddedTerms {
				// A wildcard is attached to the end of every word
				// regardless of whether or not a wildcard exists
				if !strings.HasSuffix(term, "*") {
					wildcardAddedTerms[index] = term + ":*"
				}
			}
```

agarciamontoro commented on store/sqlstore/file_info_store.go:
```diff
@@ -648,21 +675,39 @@ func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, team
+
+			wildcardAddedTerms := strings.Fields(terms)
+
+			// A wildcard is attached to the end of every word
+			// regardless of whether or not a wildcard exists
+			for index, term := range wildcardAddedTerms {
+				wildcardAddedTerms[index] = wildcardRegExpForFileSearch.ReplaceAllLiteralString(term, "*")
+			}
```
Same here
```suggestion
			// A wildcard is attached to the end of every word
			// regardless of whether or not a wildcard exists
			for index, term := range wildcardAddedTerms {
				if !strings.HasSuffix(term, "*") {
					wildcardAddedTerms[index] = term + "*"
				}
			}
```

agarciamontoro reviewed (changes requested):
Sorry for the delay on this, @kyeongsoosoo. I've added a few more suggestions to reword the comments, and also some questions in a couple of places to try and avoid using regular expressions wherever we can. Let me know if my suggestions make sense, since the logic is a bit complex and I'm sure you have it more clear in your head!

Again, sorry for the delay, my bad!

agarciamontoro commented:
Also, asking for Ibrahim's review again, since his review was removed by mistake. I need more eyes on this! :)

kyeongsoosoo commented:
@agarciamontoro Thank you for the review. I'll work on it this week.

kyeongsoosoo commented:
@agarciamontoro Sorry about the delay. I couldn't work on this PR because I was busy for a while. Would it be okay to start over from now?

M-ZubairAhmed commented:
> @agarciamontoro Sorry about the delay. I couldn't work on this PR because I was busy for a while. Would it be okay to start over from now?

Yes please do it. :)

agarciamontoro commented:
Sure, go ahead!

kyeongsoosoo commented:
@agarciamontoro I changed the code so that regex doesn't get complicated. Also added comments related to this change. (4461f50b66480aac66cfec0275fec2191b6353a3)

I used rebase and force-push in applying changes related to Monorepo.

isacikgoz commented on server/channels/store/searchtest/file_info_layer.go:
```diff
@@ -1185,6 +1190,67 @@ func testFileInfoSupportWildcards(t *testing.T, th *SearchTestHelper) {
+
+	t.Run("support wildcard search with another term placed after without wildcards", func(t *testing.T) {
+		params := &model.SearchParams{
+			Terms: "sear post",
```
Behaviourally speaking; I'm unsure whether we should treat this as a `sear* post`. I understand adding wildcard to the end of the search term but this didn't sound right to me.

kyeongsoosoo replied:
I think it's awkward, too. I removed the test case and fixed lint.

isacikgoz reviewed (commented):
Thanks for the contribution @kyeongsoosoo and apologies for the late review, I just have a concern over applying the wildcard but generally this looks okay to me. After we resolve the discussion we can update the PR and start doing some testing.

agarciamontoro commented on server/public/model/search_params.go:
```diff
@@ -109,6 +114,32 @@ func (p *SearchParams) GetExcludedDateMillis() (int64, int64) {
 	return GetStartOfDayMillis(date, p.TimeZoneOffset), GetEndOfDayMillis(date, p.TimeZoneOffset)
 }
 
+// Returns only the quoted terms in Terms
+func (p *SearchParams) GetExactPhraseTerms() string {
+	exactPhraseTerms := strings.Join(exactPhraseRegExpForFileSearch.FindAllString(p.Terms, -1), " ")
+
+	return exactPhraseTerms
+}
+
+// Returns only the non-quoted terms in Terms with an asterisk character at the end
+func (p *SearchParams) GetWildcardAddedNormalTerms() string {
```
Just a nit:
```suggestion
// GetExactPhraseTerms returns a space-separated string with only the quoted terms in Terms.
// For example: '"one" two "three" four' returns '"one" "three"'
func (p *SearchParams) GetExactPhraseTerms() string {
	exactPhraseTerms := strings.Join(exactPhraseRegExpForFileSearch.FindAllString(p.Terms, -1), " ")

	return exactPhraseTerms
}

// GetWildcardAddedNormalTerms returns a space-separated string with only the non-quoted
// terms in Terms, with an additional asterisk character at the end.
// For example: '"one" two "three" four' returns 'two* four*'
func (p *SearchParams) GetWildcardAddedNormalTerms() string {
```

kyeongsoosoo replied:
Thanks! I updated the documentation😄

agarciamontoro reviewed (approved):
This is great :tada: I'd love for someone from QA to give it a thorough pass, but apart from that, it looks good to me! :) Just left a minor non-blocking comment below.

Thank you for your infinite patience, @kyeongsoosoo!

kyeongsoosoo commented:
@isacikgoz 
While investigating a failing test [testFileInfoNotSupportPrecedingWildcards](https://github.com/mattermost/mattermost/blob/7b164d4f822a5b04dbf496eb55e226ec3a952217/server/channels/store/searchtest/file_info_layer.go#L1188), I found something confusing.

When using Bleve Search Engine, the preceding wildcard cannot be used alone. However, if the trailing wildcard is also used, the preceding wildcard can also be applied.
https://github.com/mattermost/mattermost/blob/7b164d4f822a5b04dbf496eb55e226ec3a952217/server/platform/services/searchengine/bleveengine/search.go#L709-L718

As I appended the trailing wildcard to every search, the test started to fail.
Should I prevent preceding wildcard searches in the store layer?

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

isacikgoz commented:
/update-branch

isacikgoz commented:
@kyeongsoosoo I couldn't find the reasoning for the preceding wildcard on bleve but I think since we implicitly allow it, I think we can change the tests. @mgdelacroix do you remember the reasoning about this?

kyeongsoosoo commented:
/update-branch

kyeongsoosoo commented on server/channels/store/searchtest/file_info_layer.go:
```diff
@@ -127,7 +127,7 @@ var searchFileInfoStoreTests = []searchTest{
 		Name: "Should not support search with preceding wildcards",
 		Fn:   testFileInfoNotSupportPrecedingWildcards,
-		Tags: []string{EngineAll},
+		Tags: []string{EnginePostgres, EngineMySql, EngineElasticSearch},
```
As the preceding wildcard is implicitly allowed on `BleveSearch`, I removed the `EngineBleve` tag from the tags.

kyeongsoosoo commented:
/update-branch

kyeongsoosoo commented:
@isacikgoz All tests now pass.

mgdelacroix commented:
> I couldn't find the reasoning for the preceding wildcard on bleve

IIRC we applied it to try and make the behavior of all engines consistent. If we want to allow them now, that's fine as the test spec works as a description of the expectations. I'd make sure that there is a test that confirms Bleve's behavior tho, besides removing the tag from the `testFileInfoNotSupportPrecedingWildcards` one

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

isacikgoz commented:
/update-branch

isacikgoz reviewed (approved):
This looks good to me, thanks for the patience @kyeongsoosoo 💯 

PS: for the CI failure; we introduced a new linter rule for empty spaces. This is good to go once it is fixed. Basically you'll need to remove those empty lines :)

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

kyeongsoosoo commented:
/update-branch

kyeongsoosoo commented:
@isacikgoz  I fixed the lint issue :)

isacikgoz commented:
/update-branch

hanzei commented:
/update-branch

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

hanzei commented:
@isacikgoz @agarciamontoro The changes broke a test in the EE code:
```
=== FAIL: github.com/mattermost/enterprise/elasticsearch EE/PATH/TestSearchFileInfoStore/Should_always_be_able_to_support_wildcard_searches/support_wildcard_search_only_for_plain_terms_when_plain_terms_and_quoted_terms_are_used_together. (0.00s)
    file_info_layer.go:1229: 
        	Error Trace:	/enterprise/mattermost/server/channels/store/searchtest/file_info_layer.go:1229
        	Error:      	"map[]" should have 1 item(s), but has 0
        	Test:       	EE/PATH/TestSearchFileInfoStore/Should_always_be_able_to_support_wildcard_searches/support_wildcard_search_only_for_plain_terms_when_plain_terms_and_quoted_terms_are_used_together.
```

Would you of you be open on looking into the broken test?
